daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala

```diff
@@ -657,6 +668,64 @@ trait DataOutputStreamImplMixin extends DataStreamCommonState
     }
   }
 
+  final def putFile(
+    path: Path,
+    lengthInBits: Long,
+    chunkSizeInBytes: Int,
+    finfo: FormatInfo): Long = {
+
+    val fileStream =
+      try {
+        Files.newInputStream(path, StandardOpenOption.READ)
+      } catch {
+        case e: Exception =>
+          throw new BlobIOException(
+            "Unable to open BLOB %s for reading: %s".format(
+              path.toString,
+              e.getMessage()))
+      }
+
+    val array = new Array[Byte](chunkSizeInBytes)
+
+    val chunkSizeInBits = chunkSizeInBytes * 8
+
+    var remainingBitsToPut = lengthInBits
+    var fileHasData = true
+    var bitsWritten: Long = 0
+    while (remainingBitsToPut > 0 && fileHasData) {
+      val bitsToRead = Math.min(remainingBitsToPut, chunkSizeInBits)
+      val bytesToRead = (bitsToRead + 7) / 8
+      val bytesRead = fileStream.read(array, 0, bytesToRead.toInt)
+      if (bytesRead == -1) {
+        fileHasData = false
+      } else {
+        val bitsToPut = Math.min(bytesRead * 8, bitsToRead)
+        val ret = putByteArray(array, bitsToPut.toInt, finfo, ignoreByteOrder=true)
+        if (!ret) {
+          fileStream.close()
+          throw new BlobIOException(
+            "Failed to write BLOB data: %s".format(path.toString))
+        }
+        bitsWritten += bitsToPut
+        remainingBitsToPut -= bitsToPut
+      }
+    }
+
+    fileStream.close()
+
+    // calculate the skip bits
+    val nFillBits = remainingBitsToPut
+    if (nFillBits > 0) {
+      val ret = skip(nFillBits, finfo)
+      if (!ret) {
+          throw new BlobIOException(
```

Comment:
How can this fail? 

Can this skip call if nFillBits is enough to cause any existing fragment byte to be filled out, resulting in writing 1 more byte to the output, which could fail if the file-system is full? 

Or are we re-establishing the fragment byte here? 

If the latter, then this can't fail for any user-sensible reason, so should  just be a Assert.invariant(ret) to insist the skip call succeeds.